src/bin/upgrade/main.rs

```diff
@@ -87,58 +98,98 @@ fn update_manifest(
             if (only_update.is_empty() || only_update.contains(name)) &&
                 is_version_dependency(old_value)
             {
-                let latest_version = get_latest_dependency(name, allow_prerelease)?;
+                let latest_version = new_dependency(name)?;
 
                 manifest.update_table_entry(&table_path, &latest_version)?;
             }
         }
     }
 
     let mut file = Manifest::find_file(&manifest_path)?;
-    manifest.write_to_file(&mut file)?;
+    manifest
+        .write_to_file(&mut file)
+        .chain_err(|| "Failed to write new manifest contents")
+}
 
-    Ok(())
+fn upgrade_manifest(
+    manifest_path: &Option<String>,
+    only_update: &[String],
+    allow_prerelease: bool,
+) -> Result<()> {
+    upgrade_manifest_using_dependencies(manifest_path, only_update, |name| {
+        get_latest_dependency(name, allow_prerelease)
+    })
+}
+
+fn upgrade_manifest_from_cache(
+    manifest_path: &Option<String>,
+    only_update: &[String],
+    new_deps: &HashMap<String, Dependency>,
+) -> Result<()> {
+    upgrade_manifest_using_dependencies(
+        manifest_path,
+        only_update,
+        |name| Ok(new_deps[name].clone()),
+    )
 }
 
 /// Get a list of the paths of all the (non-virtual) manifests in the workspace.
 fn get_workspace_manifests(manifest_path: &Option<String>) -> Result<Vec<String>> {
-    let mut metadata_gatherer = Command::new("cargo");
-    metadata_gatherer.args(&["metadata", "--no-deps", "--format-version", "1", "-q"]);
+    Ok(
+        cargo_metadata::metadata_deps(manifest_path.as_ref().map(|p| Path::new(p)), true)
+            .chain_err(|| "Failed to get metadata")?
+            .packages
+            .iter()
+            .map(|p| p.manifest_path.clone())
+            .collect(),
+    )
+}
 
-    if let Some(ref manifest_path) = *manifest_path {
-        metadata_gatherer.args(&["--manifest-path", manifest_path]);
-    }
+/// Look up all current direct crates.io dependencies in the workspace. Then get the latest version
+/// for each.
+fn get_new_workspace_deps(
+    manifest_path: &Option<String>,
+    only_update: &[String],
+    allow_prerelease: bool,
+) -> Result<HashMap<String, Dependency>> {
+    let mut new_deps = HashMap::new();
+
+    cargo_metadata::metadata_deps(manifest_path.as_ref().map(|p| Path::new(p)), true)
+        .chain_err(|| "Failed to get metadata")?
+        .packages
+        .iter()
+        .flat_map(|package| package.dependencies.to_owned())
+        .filter(|dependency| {
+            only_update.is_empty() || only_update.contains(&dependency.name)
+        })
+        .map(|dependency| {
+            if !new_deps.contains_key(&dependency.name) {
```

Comment:
I'd probably split that into filter + map

Reply:
Having the `if` explicit makes the order of evaluation more obvious (rather than implicitly relying on lazy evaluation).

I don't feel too strongly, though, so will change it accordingly.

Reply:
Ah. Moving to filter + map is problematic. The filer has an immutable borrow of `new_deps` which conflicts with the mutable borrow in the map.

I'll leave as-is, then.

Reply:
Ah too bad. Okay, let's keep it then. The rest is good as-is as well.